### PR TITLE
[SPARK-49666][SQL] Add feature flag for trim collation feature

### PR DESCRIPTION
**Pull request by jovanpavl-db**
### What changes were proposed in this pull request?
Introducing new specifier for trim collations (both leading and trailing trimming). These are initial changes so that trim specifier is recognized and put under feature flag (all code paths blocked).

### Why are the changes needed?
Support for trailing space trimming is one of the requested feature by users.

### Does this PR introduce _any_ user-facing change?
This is guarded by feature flag. 

### How was this patch tested?
Added tests to CollationSuite, SqlConfSuite and QueryCompilationErrorSuite.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by stefankandic — CHANGES_REQUESTED** [review]
Nice work! Left some minor comments but looks good overall

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -2557,6 +2557,13 @@ class AstBuilder extends DataTypeAstBuilder
   }
 
   override def visitCollateClause(ctx: CollateClauseContext): String = withOrigin(ctx) {
+    val collationName = ctx.collationName.getText
+    if (!SQLConf.get.trimCollationEnabled &&
+      (collationName.toUpperCase().contains("TRIM") ||
```
[the remark this answers is not in the document]
Comment (jovanpavl-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2562]: Correct, my bad.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -759,6 +759,14 @@ object SQLConf {
       .checkValue(_ > 0, "The initial number of partitions must be positive.")
       .createOptional
 
+  lazy val TRIM_COLLATION_ENABLED =
+    buildConf("spark.sql.trimCollation.enabled")
```
[the remark this answers is not in the document]
Comment (jovanpavl-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Accepted.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 3]
```diff
@@ -18,15 +18,13 @@
 package org.apache.spark.sql.execution.datasources
 
 import java.util.Locale
-
```
[the remark this answers is not in the document]
Comment (jovanpavl-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: reverted.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 4]
```diff
@@ -661,58 +775,111 @@ private static int collationNameToId(
...
+        CaseSensitivity caseSensitivity = CaseSensitivity.CS;         // Default: Case Sensitive
+        AccentSensitivity accentSensitivity = AccentSensitivity.AS;    // Default: Accent Sensitive
+        SpaceTrimming spaceTrimming = SpaceTrimming.NONE;          // Default: No Trim
+
+        String[] specifiers = remainingSpecifiers.split("_");
+
+        // Iterate through specifiers and set corresponding flags
+        for (String specifier : specifiers) {
+          switch (specifier) {
+            case "CI":
```
[the remark this answers is not in the document]
Comment (jovanpavl-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:808]: Very nice catch, thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 5]
```diff
@@ -640,6 +639,26 @@ case class QualifyLocationWithWarehouse(catalog: SessionCatalog) extends Rule[Lo
   }
 }
 
+object TrimCollationCheck extends (LogicalPlan => Unit) {
```
[the remark this answers is not in the document]
Comment (jovanpavl-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala:644]: It's not enough for the statements like: `SELECT collate('aaa', 'UNICODE_TRIM')`.  Here default collation is UTF8_BINARY so changes in SQLConf won't help.

[the remark this answers is not in the document]
Comment (stefankandic) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala:644]: What if we add the check also in the `CollateExpressionBuilder`?
https://github.com/apache/spark/blob/587cd2b18c37b35f641320eb176d7638b2e52b42/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collationExpressions.scala#L55

[the remark this answers is not in the document]
Comment (jovanpavl-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: Yes, that would work. Done!
For my own knowledge is there a reason we avoid adding analyzer rules?

[the remark this answers is not in the document]
Comment (stefankandic) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: Analyzer is pretty slow right now, it runs the rules iteratively over and over again until the plan gets to the fixed point - which can be very slow for bigger queries. That's why I try to avoid introducing anything new that could increase that complexity further.

Btw, my analyzer knowledge is also very shallow, maybe @vladimirg-db can provide an expert opinion here 😄

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 6]
```diff
@@ -759,6 +759,14 @@ object SQLConf {
       .checkValue(_ > 0, "The initial number of partitions must be positive.")
       .createOptional
 
+  lazy val TRIM_COLLATION_ENABLED =
+    buildConf("spark.sql.collation.trim.enabled")
+      .doc("Trim collation feature is under development and its use should be done under this" +
```
Comment (vladimirg-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:764]: Would be nice to have a description for this feature here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (vladimirg-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: Sure.
- The rules run in fixed point (many iterations) over the whole logical plan, which can be super slow for big logical plans. Every new rule slows down those corner cases.
- Analyzer logic depends on the ordering of rules, which is very hard to get right, and there are frequent regressions related to unobvious outcome of some new rule.

**Review by stefankandic — APPROVED** [review]
LGTM pending scalastyle fixes!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 6 above)
Comment (jovanpavl-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Done.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 7]
```diff
@@ -200,6 +201,7 @@ public Collation(
      * bit 28-24: Reserved.
      * bit 23-22: Reserved for version.
      * bit 21-18: Reserved for space trimming.
+     * 0000 = none, 0001 = left trim, 0010 = right trim, 0011 = trim.
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:204]: here's a thought: should we consider having one bit signaling whether there is any trimming sensitivity or not?
while I would agree that it's not immediately useful, it might come in handy in the future

thoughts? cc @nikolamand-db @stefankandic

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 8]
```diff
@@ -259,6 +269,14 @@ protected enum ImplementationProvider {
         UTF8_BINARY, ICU
       }
 
+      /**
+       * Bits 19-18 having value 00 for no space trimming, 01 for left space trimming
+       * 10 for right space trimming and 11 for both sides space trimming.
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:274]: if we end up going with this, I would just add a note in this comment that 1 bit still remains reserved at this time
that is, if we don't go with: https://github.com/apache/spark/pull/48222/files#r1774931595

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java [hunk 9]
```diff
@@ -464,29 +548,52 @@ protected Collation buildCollation() {
...
+       * - Base collation name (UTF8_BINARY or UTF8_LCASE)
+       * - Optional space trimming when non-default preceded by underscore
+       * Examples: UTF8_BINARY, UTF8_BINARY_LCASE_LTRIM, UTF8_BINARY_TRIM.
+       */
+      @Override
+      protected String normalizedCollationName() {
+        StringBuilder builder = new StringBuilder();
+        if(caseSensitivity == CaseSensitivity.UNSPECIFIED){
+          builder.append(UTF8_BINARY_COLLATION_NAME);
+        }else{
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:589]:
```suggestion
        } else{
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java (hunk 7 above)
Comment (jovanpavl-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:204]: Can you provide some use-case where this one bit would be useful? IMO let's leave one bit reserved, maybe for different class of trimmings (for example not just ascii 20).

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java (hunk 8 above)
Comment (jovanpavl-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java:274]: Sure, will update.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 10]
```diff
@@ -77,9 +107,15 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
 
   test("collate function syntax with default collation set") {
     withSQLConf(SqlApiConf.DEFAULT_COLLATION -> "UTF8_LCASE") {
-      assert(sql(s"select collate('aaa', 'utf8_lcase')").schema(0).dataType ==
-        StringType("UTF8_LCASE"))
+      assert(
+        sql(s"select collate('aaa', 'utf8_lcase')").schema(0).dataType ==
+        StringType("UTF8_LCASE")
+      )
```
Comment (uros-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:113]: we should avoid unnecessary changes

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 11]
```diff
@@ -44,27 +44,57 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
...
-        == StringType(collationId))
+      assert(
+        sql(s"select 'aaa' collate $collationName").schema(0).dataType
+        == StringType(collationId)
+      )
     }
   }
 
   test("collation expression returns name of collation") {
-    Seq("utf8_binary", "utf8_lcase", "unicode", "unicode_ci").foreach { collationName =>
```
Comment (uros-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:65]: top avoid repetition across tests, let's consider using something like `testSuppCollations`

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala (hunk 10 above)
Comment (jovanpavl-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:113]: I got skalastyle failing without clear message in which file (bad developer expirience). So what I had to do is go through all methods I changed (or added) and do formatting by scalastyle..

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 12]
```diff
@@ -149,6 +149,10 @@ class SparkSqlAstBuilder extends AstBuilder {
    * }}}
    */
   override def visitSetCollation(ctx: SetCollationContext): LogicalPlan = withOrigin(ctx) {
+    val collationName = ctx.collationName.getText
+    if (!SQLConf.get.trimCollationEnabled && collationName.toUpperCase().contains("TRIM")) {
+      throw QueryCompilationErrors.trimCollationNotEnabledError()
+    }
```
Comment (uros-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:155]: do we have a test for this? i.e. e2e sql test with "SET COLLATION ..._trim;")

Comment (uros-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:155]: also, given the changes in "sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala", are we sure that we need this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collationExpressions.scala
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collationExpressions.scala:1]: iiuc, this is for Collation

what about Collate?

File: common/utils/src/main/resources/error/error-conditions.json [hunk 13]
```diff
@@ -4886,11 +4886,6 @@
           "Catalog <catalogName> does not support <operation>."
         ]
       },
-      "COLLATION" : {
-        "message" : [
-          "Collation is not yet supported."
-        ]
-      },
```
Comment (uros-db) [inline common/utils/src/main/resources/error/error-conditions.json:4893]: why are we removing this now?

File: common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala [hunk 14]
```diff
@@ -369,8 +369,6 @@ class CollationFactorySuite extends AnyFunSuite with Matchers { // scalastyle:ig
       1 << 15, // UTF8_BINARY mandatory zero bit 15 breach.
       1 << 16, // UTF8_BINARY mandatory zero bit 16 breach.
       1 << 17, // UTF8_BINARY mandatory zero bit 17 breach.
-      1 << 18, // UTF8_BINARY mandatory zero bit 18 breach.
-      1 << 19, // UTF8_BINARY mandatory zero bit 19 breach.
```
Comment (uros-db) [inline common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala:373]: and what about bit 20? do we still keep it as mandatory zero
also, what about bit 21? I don't even see that one here...

File: common/utils/src/main/resources/error/error-conditions.json (hunk 13 above)
Comment (jovanpavl-db) [reply common/utils/src/main/resources/error/error-conditions.json:4893]: UNSUPPORTED_FEATURE.TRIM_COLLATION was removed and this is the leftover that doesn't make sense without it.

File: common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala (hunk 14 above)
Comment (jovanpavl-db) [reply common/unsafe/src/test/scala/org/apache/spark/unsafe/types/CollationFactorySuite.scala:373]: Yes we do, untill we find a reason to use it, it should be reserved and fixed to 0.
There were no test for 21, will add it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collationExpressions.scala
Comment (jovanpavl-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collationExpressions.scala:1]: Not sure I get the comment.
This covers Collate, take a look at tests I added at sql/core/src/test/scala/org/apache/spark/sql/errors/QueryCompilationErrorsSuite.scala, it covers collate clause, for example:
SELECT collate('aaa', 'UNICODE_TRIM').

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 12 above)
Comment (jovanpavl-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:155]: Nope, not the same code path, AstBuilder is not being hit.
Point about test missing is fair, my bad.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala (hunk 11 above)
Comment (jovanpavl-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:65]: I find this being better readable, however am not hard opinionated.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 15]
```diff
@@ -759,6 +759,17 @@ object SQLConf {
       .checkValue(_ > 0, "The initial number of partitions must be positive.")
       .createOptional
 
+  lazy val TRIM_COLLATION_ENABLED =
+    buildConf("spark.sql.collation.trim.enabled")
+      .doc(
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:764]: let's mark it as `.internal()`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 16]
```diff
@@ -2557,6 +2557,10 @@ class AstBuilder extends DataTypeAstBuilder
   }
 
   override def visitCollateClause(ctx: CollateClauseContext): String = withOrigin(ctx) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2559]: will the collate parser eventually invoke `CollateExpressionBuilder`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 15 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:764]: +1

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 17]
```diff
@@ -44,27 +44,57 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
...
   }
 
   test("collation name is case insensitive") {
-    Seq("uTf8_BiNaRy", "utf8_lcase", "uNicOde", "UNICODE_ci").foreach { collationName =>
+    Seq(
+      "uTf8_BiNaRy",
+      "utf8_lcase",
+      "uNicOde",
+      "UNICODE_ci",
+      "uNiCoDE_ltRIm_cI",
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:71]: so trim can be either in the middle or in the end?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 15 above)
Comment (jovanpavl-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:764]: Pending.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 16 above)
Comment (jovanpavl-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2559]: No, for statements like: SELECT 'aaa' COLLATE UNICODE_LTRIM_CI it won't happen.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
